File: CyborgPunch/CyborgPunch/CyborgPunch/Core/Transform.cs
using Microsoft.Xna.Framework;
using System.Collections.Generic;

namespace CyborgPunch.Core
{
    public class Transform : Component
    {
        private List<Transform> children;
        private HashSet<int> childrenHash;

        private Transform parent;
        public Transform Parent
        {
            get { return this.parent; }
            set
            {
                if (value == this)
                    return;

                //handle previous parent
                if (this.parent != null)
                {
                    this.parent.RemoveChild(this);
                }

                //setup new parent relationship
                this.parent = value;
                if (this.parent != null)
                {
                    this.parent.AddChild(this);
                }


                DetermineLocalPosition();
            }
        }

        private Vector2 position;
        public Vector2 Position
        {
            get { return this.position; }
            set { 
                this.position = value;
                DetermineLocalPosition();
            }
        }

        private Vector2 localPosition;
        public Vector2 LocalPosition
        {
            get { return this.localPosition; }
            set 
            { 
                this.localPosition = value;
                DeterminePosition();
            }
        }

        private void AddChild(Transform child)
        {
            int id = child.blob.ID;
            if (childrenHash.Add(id))
            {
                children.Add(child);
                child.DetermineLocalPosition();
            }
        }

        private void RemoveChild(Transform child)
        {
            int id = child.blob.ID;
            if (childrenHash.Contains(id))
            {
                childrenHash.Remove(id);
                children.Remove(child);
            }
        }

        void DeterminePosition()
        {
            position = localPosition;
            Transform nextParent = parent;
            while (nextParent != null)
            {
                position += nextParent.localPosition;
                nextParent = nextParent.parent;
            }

            RefreshChildren();
        }

        void DetermineLocalPosition()
        {
            localPosition = position;
            Transform nextParent = parent;
            while (nextParent != null)
            {
                localPosition -= nextParent.localPosition;
                nextParent = nextParent.parent;
            }

            RefreshChildren();
        }

        void RefreshChildren()
        {
            for (int i = 0; i < children.Count; i++)
            {
                children[i].DeterminePosition();
            }
        }

        public Transform()
        {
            children = new List<Transform>();
            childrenHash = new HashSet<int>();
            Position = Vector2.Zero;
        }

        public void Translate(float x, float y)
        {
            this.LocalPosition += new Vector2(x, y);
        }

        public void Translate(Vector2 change)
        {
            this.LocalPosition += change;
        }
    }
}

File: trunk/PunchHarder/trunk/PunchHarder/engine/transform .cpp
#include "transform.h"
#include "engine.h"

/////////////////////
// Constructors
/////////////////////
Transform::Transform()
{
    m_pParent = NULL;
    m_Position = sf::Vector3f();
    m_LocalPosition = sf::Vector3f();
}

Transform::~Transform()
{
}

/////////////////////
// Public Functions
/////////////////////

sf::Vector3f Transform::GetPosition()
{
    return m_Position;
}

void Transform::SetPosition(sf::Vector3f pos)
{
    m_Position = pos;
    DetermineLocalPosition();
}

sf::Vector3f Transform::GetLocalPosition()
{
    return m_LocalPosition;
}

void Transform::SetLocalPosition(sf::Vector3f pos)
{
    m_LocalPosition = pos;
    DeterminePosition();
}

std::list<Transform*> Transform::GetChildren()
{
    return m_Children;
}

Transform* Transform::GetParent()
{
    return m_pParent;
}

// TODO handle parent-child loop
void Transform::SetParent(Transform* parent)
{
    if(parent == this)
    {
        Debug::Log("Parent == this");
        return;
    }

    if(parent == m_pParent)
    {
        return;
    }

    //handle the previous parent
    if(m_pParent != NULL)
    {
        m_pParent->RemoveChild(this);
    }

    //setup new parent relationship
    if(parent == NULL)
    {
        //removing my parent, add to root node
        GameObject* root = GameObjectManager::GetInstance()->GetRoot();
        m_pParent = root->GetComponent<Transform>();
    }
    else
    {
        //adding a parent, we are now a child
        m_pParent = parent;
        m_pParent->AddChild(this);
    }

    DetermineLocalPosition();
}

void Transform::Translate(float x, float y, float z)
{
    SetLocalPosition(m_LocalPosition + sf::Vector3f(x,y,z));
}

void Transform::Translate(sf::Vector3f delta)
{
    SetLocalPosition(m_LocalPosition + delta);
}

/////////////////////
// Virtual Functions
/////////////////////
void Transform::Update()
{
    //call base class function
    Component::Update();
}

void Transform::Draw()
{
    //call base class function
    Component::Draw();
}

/////////////////////
// Private Functions
/////////////////////

void Transform::RemoveChild(Transform* child)
{
    TransformIterator itr = m_Children.begin();
    for(; itr != m_Children.end(); itr++)
    {
        Transform* current = *itr;
        if(current == child)
        {
            m_Children.erase(itr);
            return;
        }
    }

    Debug::Log("Child was never found : " + child->ToString());
}

void Transform::AddChild(Transform* child)
{
    m_Children.push_back(child);
    child->DetermineLocalPosition();
}

void Transform::DeterminePosition()
{
    m_Position = m_LocalPosition;
    Transform* pNextParent = m_pParent;
    while(pNextParent != NULL)
    {
        m_Position += pNextParent->GetLocalPosition();
        pNextParent = pNextParent->GetParent();
    }

    RefreshChildren();
}

void Transform::DetermineLocalPosition()
{
    m_LocalPosition = m_Position;
    Transform* pNextParent = m_pParent;
    while(pNextParent != NULL)
    {
        m_LocalPosition -= pNextParent->GetLocalPosition();
        pNextParent = pNextParent->GetParent();
    }

    RefreshChildren();
}

void Transform::RefreshChildren()
{
    TransformIterator itr = m_Children.begin();
    for(; itr != m_Children.end(); itr++)
    {
        Transform* current = *itr;
        current->DeterminePosition();
    }
}
